### ISSUE TITLE
Race condition in shared form mapping resolution logic

### ISSUE THREAD

**Issue by captainsafia** [dotnet/aspnetcore#50464]
This behavior can be reproed with the benchmarks available at https://github.com/aspnet/Benchmarks/pull/1900.

Repro:

```
crank --config ./src/BenchmarksApps/Mvc/benchmarks.mvcjson.yml --scenario MapActionEchoTodoFormMapping --profile aspnet-perf-lin --profile short --application.framework net8.0 --application.source.branchOrCommit safia/fromform-tes
```

Exceptions:
```
fail: Microsoft.AspNetCore.Server.Kestrel[13]
      Connection id "0HMTAABU203IC", Request id "0HMTAABU203IC:00000001": An unhandled exception was thrown by the application.
      System.InvalidOperationException: Cannot end a resolution graph while one is not in progress.
         at Microsoft.AspNetCore.Components.Endpoints.FormMapping.Metadata.FormDataMetadataFactory.FormMetadataContext.EndResolveGraph()
         at Microsoft.AspNetCore.Components.Endpoints.FormMapping.Metadata.FormDataMetadataFactory.GetOrCreateMetadataFor(Type type, FormDataMapperOptions options)
         at Microsoft.AspNetCore.Components.Endpoints.FormMapping.ComplexTypeConverterFactory.CanConvert(Type type, FormDataMapperOptions options)
         at Microsoft.AspNetCore.Components.Endpoints.FormMapping.FormDataMapperOptions.CreateConverter(Type type, FormDataMapperOptions options)
         at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd[TArg](TKey key, Func`3 valueFactory, TArg factoryArgument)
         at Microsoft.AspNetCore.Components.Endpoints.FormMapping.FormDataMapperOptions.ResolveConverter[T]()
         at lambda_method3(Closure, Object, HttpContext, Object)
         at Microsoft.AspNetCore.Http.RequestDelegateFactory.<>c__DisplayClass103_2.<<HandleRequestBodyAndCompileRequestDelegateForForm>b__2>d.MoveNext()
      --- End of stack trace from previous location ---
         at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests[TContext](IHttpApplication`1 application)
fail: Microsoft.AspNetCore.Server.Kestrel[13]
      Connection id "0HMTAABU203IE", Request id "0HMTAABU203IE:00000001": An unhandled exception was thrown by the application.
      System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection. (Parameter 'index')
         at System.Collections.Generic.List`1.RemoveAt(Int32 index)
         at System.Collections.Generic.List`1.Remove(T item)
         at Microsoft.AspNetCore.Components.Endpoints.FormMapping.Metadata.FormDataMetadataFactory.GetOrCreateMetadataFor(Type type, FormDataMapperOptions options)
         at Microsoft.AspNetCore.Components.Endpoints.FormMapping.ComplexTypeConverterFactory.CanConvert(Type type, FormDataMapperOptions options)
         at Microsoft.AspNetCore.Components.Endpoints.FormMapping.FormDataMapperOptions.CreateConverter(Type type, FormDataMapperOptions options)
         at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd[TArg](TKey key, Func`3 valueFactory, TArg factoryArgument)
         at Microsoft.AspNetCore.Components.Endpoints.FormMapping.FormDataMapperOptions.ResolveConverter[T]()
         at lambda_method3(Closure, Object, HttpContext, Object)
         at Microsoft.AspNetCore.Http.RequestDelegateFactory.<>c__DisplayClass103_2.<<HandleRequestBodyAndCompileRequestDelegateForForm>b__2>d.MoveNext()
      --- End of stack trace from previous location ---
         at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests[TContext](IHttpApplication`1 application)
fail: Microsoft.AspNetCore.Server.Kestrel[13]
      Connection id "0HMTAABU203IL", Request id "0HMTAABU203IL:00000001": An unhandled exception was thrown by the application.
      System.InvalidOperationException: Cannot end a resolution graph while one is not in progress.
         at Microsoft.AspNetCore.Components.Endpoints.FormMapping.Metadata.FormDataMetadataFactory.FormMetadataContext.EndResolveGraph()
         at Microsoft.AspNetCore.Components.Endpoints.FormMapping.Metadata.FormDataMetadataFactory.GetOrCreateMetadataFor(Type type, FormDataMapperOptions options)
         at Microsoft.AspNetCore.Components.Endpoints.FormMapping.ComplexTypeConverterFactory.CanConvert(Type type, FormDataMapperOptions options)
         at Microsoft.AspNetCore.Components.Endpoints.FormMapping.FormDataMapperOptions.CreateConverter(Type type, FormDataMapperOptions options)
         at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd[TArg](TKey key, Func`3 valueFactory, TArg factoryArgument)
         at Microsoft.AspNetCore.Components.Endpoints.FormMapping.FormDataMapperOptions.ResolveConverter[T]()
         at lambda_method3(Closure, Object, HttpContext, Object)
         at Microsoft.AspNetCore.Http.RequestDelegateFactory.<>c__DisplayClass103_2.<<HandleRequestBodyAndCompileRequestDelegateForForm>b__2>d.MoveNext()
      --- End of stack trace from previous location ---
         at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests[TContext](IHttpApplication`1 application)
fail: Microsoft.AspNetCore.Server.Kestrel[13]
      Connection id "0HMTAABU203IF", Request id "0HMTAABU203IF:00000001": An unhandled exception was thrown by the application.
      System.InvalidOperationException: Cannot end a resolution graph while one is not in progress.
         at Microsoft.AspNetCore.Components.Endpoints.FormMapping.Metadata.FormDataMetadataFactory.FormMetadataContext.EndResolveGraph()
         at Microsoft.AspNetCore.Components.Endpoints.FormMapping.Metadata.FormDataMetadataFactory.GetOrCreateMetadataFor(Type type, FormDataMapperOptions options)
         at Microsoft.AspNetCore.Components.Endpoints.FormMapping.ComplexTypeConverterFactory.CanConvert(Type type, FormDataMapperOptions options)
         at Microsoft.AspNetCore.Components.Endpoints.FormMapping.FormDataMapperOptions.CreateConverter(Type type, FormDataMapperOptions options)
         at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd[TArg](TKey key, Func`3 valueFactory, TArg factoryArgument)
         at Microsoft.AspNetCore.Components.Endpoints.FormMapping.FormDataMapperOptions.ResolveConverter[T]()
         at lambda_method3(Closure, Object, HttpContext, Object)
         at Microsoft.AspNetCore.Http.RequestDelegateFactory.<>c__DisplayClass103_2.<<HandleRequestBodyAndCompileRequestDelegateForForm>b__2>d.MoveNext()
      --- End of stack trace from previous location ---
         at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests[TContext](IHttpApplication`1 application)
fail: Microsoft.AspNetCore.Server.Kestrel[13]
      Connection id "0HMTAABU203II", Request id "0HMTAABU203II:00000001": An unhandled exception was thrown by the application.
      System.InvalidOperationException: Cannot end a resolution graph while one is not in progress.
         at Microsoft.AspNetCore.Components.Endpoints.FormMapping.Metadata.FormDataMetadataFactory.FormMetadataContext.EndResolveGraph()
         at Microsoft.AspNetCore.Components.Endpoints.FormMapping.Metadata.FormDataMetadataFactory.GetOrCreateMetadataFor(Type type, FormDataMapperOptions options)
         at Microsoft.AspNetCore.Components.Endpoints.FormMapping.ComplexTypeConverterFactory.CanConvert(Type type, FormDataMapperOptions options)
         at Microsoft.AspNetCore.Components.Endpoints.FormMapping.FormDataMapperOptions.CreateConverter(Type type, FormDataMapperOptions options)
         at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd[TArg](TKey key, Func`3 valueFactory, TArg factoryArgument)
         at Microsoft.AspNetCore.Components.Endpoints.FormMapping.FormDataMapperOptions.ResolveConverter[T]()
         at lambda_method3(Closure, Object, HttpContext, Object)
         at Microsoft.AspNetCore.Http.RequestDelegateFactory.<>c__DisplayClass103_2.<<HandleRequestBodyAndCompileRequestDelegateForForm>b__2>d.MoveNext()
      --- End of stack trace from previous location ---
         at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests[TContext](IHttpApplication`1 application)
fail: Microsoft.AspNetCore.Server.Kestrel[13]
      Connection id "0HMTAABU203IH", Request id "0HMTAABU203IH:00000001": An unhandled exception was thrown by the application.
      System.InvalidOperationException: Cannot end a resolution graph while one is not in progress.
         at Microsoft.AspNetCore.Components.Endpoints.FormMapping.Metadata.FormDataMetadataFactory.FormMetadataContext.EndResolveGraph()
         at Microsoft.AspNetCore.Components.Endpoints.FormMapping.Metadata.FormDataMetadataFactory.GetOrCreateMetadataFor(Type type, FormDataMapperOptions options)
         at Microsoft.AspNetCore.Components.Endpoints.FormMapping.ComplexTypeConverterFactory.CanConvert(Type type, FormDataMapperOptions options)
         at Microsoft.AspNetCore.Components.Endpoints.FormMapping.FormDataMapperOptions.CreateConverter(Type type, FormDataMapperOptions options)
         at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd[TArg](TKey key, Func`3 valueFactory, TArg factoryArgument)
         at Microsoft.AspNetCore.Components.Endpoints.FormMapping.FormDataMapperOptions.ResolveConverter[T]()
         at lambda_method3(Closure, Object, HttpContext, Object)
         at Microsoft.AspNetCore.Http.RequestDelegateFactory.<>c__DisplayClass103_2.<<HandleRequestBodyAndCompileRequestDelegateForForm>b__2>d.MoveNext()
      --- End of stack trace from previous location ---
         at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests[TContext](IHttpApplication`1 application)
fail: Microsoft.AspNetCore.Server.Kestrel[13]
      Connection id "0HMTAABU203IM", Request id "0HMTAABU203IM:00000001": An unhandled exception was thrown by the application.
      System.InvalidOperationException: Cannot end a resolution graph while one is not in progress.
         at Microsoft.AspNetCore.Components.Endpoints.FormMapping.Metadata.FormDataMetadataFactory.FormMetadataContext.EndResolveGraph()
         at Microsoft.AspNetCore.Components.Endpoints.FormMapping.Metadata.FormDataMetadataFactory.GetOrCreateMetadataFor(Type type, FormDataMapperOptions options)
         at Microsoft.AspNetCore.Components.Endpoints.FormMapping.ComplexTypeConverterFactory.CanConvert(Type type, FormDataMapperOptions options)
         at Microsoft.AspNetCore.Components.Endpoints.FormMapping.FormDataMapperOptions.CreateConverter(Type type, FormDataMapperOptions options)
         at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd[TArg](TKey key, Func`3 valueFactory, TArg factoryArgument)
         at Microsoft.AspNetCore.Components.Endpoints.FormMapping.FormDataMapperOptions.ResolveConverter[T]()
         at lambda_method3(Closure, Object, HttpContext, Object)
         at Microsoft.AspNetCore.Http.RequestDelegateFactory.<>c__DisplayClass103_2.<<HandleRequestBodyAndCompileRequestDelegateForForm>b__2>d.MoveNext()
      --- End of stack trace from previous location ---
         at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests[TContext](IHttpApplication`1 application)
fail: Microsoft.AspNetCore.Server.Kestrel[13]
      Connection id "0HMTAABU203IG", Request id "0HMTAABU203IG:00000001": An unhandled exception was thrown by the application.
      System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection. (Parameter 'index')
         at System.Collections.Generic.List`1.RemoveAt(Int32 index)
         at System.Collections.Generic.List`1.Remove(T item)
         at Microsoft.AspNetCore.Components.Endpoints.FormMapping.Metadata.FormDataMetadataFactory.GetOrCreateMetadataFor(Type type, FormDataMapperOptions options)
         at Microsoft.AspNetCore.Components.Endpoints.FormMapping.ComplexTypeConverterFactory.CanConvert(Type type, FormDataMapperOptions options)
         at Microsoft.AspNetCore.Components.Endpoints.FormMapping.FormDataMapperOptions.CreateConverter(Type type, FormDataMapperOptions options)
         at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd[TArg](TKey key, Func`3 valueFactory, TArg factoryArgument)
         at Microsoft.AspNetCore.Components.Endpoints.FormMapping.FormDataMapperOptions.ResolveConverter[T]()
         at lambda_method3(Closure, Object, HttpContext, Object)
         at Microsoft.AspNetCore.Http.RequestDelegateFactory.<>c__DisplayClass103_2.<<HandleRequestBodyAndCompileRequestDelegateForForm>b__2>d.MoveNext()
      --- End of stack trace from previous location ---
         at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests[TContext](IHttpApplication`1 application)
fail: Microsoft.AspNetCore.Server.Kestrel[13]
      Connection id "0HMTAABU203ID", Request id "0HMTAABU203ID:00000001": An unhandled exception was thrown by the application.
      System.InvalidOperationException: Cannot begin a new resolution graph while one is already in progress.
         at Microsoft.AspNetCore.Components.Endpoints.FormMapping.Metadata.FormDataMetadataFactory.FormMetadataContext.BeginResolveGraph()
         at Microsoft.AspNetCore.Components.Endpoints.FormMapping.Metadata.FormDataMetadataFactory.GetOrCreateMetadataFor(Type type, FormDataMapperOptions options)
         at Microsoft.AspNetCore.Components.Endpoints.FormMapping.ComplexTypeConverterFactory.CanConvert(Type type, FormDataMapperOptions options)
         at Microsoft.AspNetCore.Components.Endpoints.FormMapping.FormDataMapperOptions.CreateConverter(Type type, FormDataMapperOptions options)
         at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd[TArg](TKey key, Func`3 valueFactory, TArg factoryArgument)
         at Microsoft.AspNetCore.Components.Endpoints.FormMapping.FormDataMapperOptions.ResolveConverter[T]()
         at lambda_method3(Closure, Object, HttpContext, Object)
         at Microsoft.AspNetCore.Http.RequestDelegateFactory.<>c__DisplayClass103_2.<<HandleRequestBodyAndCompileRequestDelegateForForm>b__2>d.MoveNext()
      --- End of stack trace from previous location ---
         at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests[TContext](IHttpApplication`1 application)
info: Microsoft.Hosting.Lifetime[0]
      Application is shutting down...
```

**Comment by javiercn** [comment]
@captainsafia good catch. 

Only the `add` part is "atomic" on the call to `GetOrAdd` in the concurrent dictionary, so the shared state there causes issues.

**Comment by javiercn** [comment]
I've added a candidate fix [here](903ef902c929a1cd1b173355c07520911d444b5e)

**Comment by captainsafia** [comment]
> I've added a candidate fix [here](https://github.com/dotnet/aspnetcore/commit/903ef902c929a1cd1b173355c07520911d444b5e)

Thanks! Looks like the fix does this trick.
